src/monitor/group_state_machine.c

```diff
@@ -1608,8 +1603,7 @@ IsUnhealthy(AutoFailoverNode *pgAutoFailoverNode)
 	 * If the keeper reports that PostgreSQL is not running, then the node
 	 * isn't Healthy.
 	 */
-	if (!pgAutoFailoverNode->pgIsRunning &&
-		IsStateIn(pgAutoFailoverNode->goalState, pgIsNotRunningStateList))
+	if (!pgAutoFailoverNode->pgIsRunning)
```

Comment:
I remember that we added these states in MS [PR](https://github.com/citusdata/pg_auto_failover/pull/326/commits/f91be34d0300455b41aa1ab54a02e8cd231b4e71) to prevent some of the edge cases. Given that the PR is pretty large, I cannot easily remember what the reason  was.

I'm mostly skeptical about `REPLICATION_STATE_PREPARE_MAINTENANCE `.  Do you remember any edge cases around it?

Reply:
Well the idea with having a list of states in the `IsUnhealthy` function would be to implement exceptions where we could have the information that `pgIsRunning` is false and yet would not consider the node as UnHealthy.

I think that in the current version of the Group FSM code, we don't depend on the value of pgIsRunning or the result of the IsUnhealthy function call result when we are in a state allowing Postgres to be down. This will impact a couple code path in the multiple standby (s)election mechanism, where we skip a node when it is unhealthy. That said, we don't want to target a node that is in one of the listed states anyway, so we're good.

Reply:
Some extra context on this. It seems that real bug is that the previous logic was simply incorrect, i.e. it's missing a `!` before `IsStateIn`. Instead of reporting unhealthy when the node is not running and it's in a state where it should be running, it reports unhealthy when it's in a state where it should not be running. 

This already seems the case in the commit linked by @onderkalaci. So I don't see how this logic ever worked as expected. This makes me think that it's fine to completely remove it for now.